ProductInfo에 필요한 네트워크 관련 작업

EnnzoKim opened the pull request:
## Screenshots 📸

|Name|
|:-:|
|(image here)|

<br/><br/>

## 고민, 과정, 근거 💬

- `ProductInfo`에 필요한 네트워크 관련 모듈 및 파일을 구현했습니다.
- @WhiteHyun 님이 이미 잘 구현해주셔서 거의 `HomeAPI` 관련 코드와 동일합니다.
- PR이 너무 커질 것을 우려하여, View로 표시되는 작업은 다음 PR에서 작업할 예정입니다.

## 개인적으로 느끼는 현재 문제
> 제가 `ProductInfo`관련한 네트워크 작업을 하면서 느꼈던 문제입니다.
섣불리 스스로 고치기 전에 의견을 공유하는 것이 먼저라고 판단하여, 고친 것은 아직 아무것도 없습니다!

- 현재 Scene별로 API가 모듈화 되어있습니다. 
`ProductInfo` 상세정보와 `Home`의 상품의 리스트와 정보가 거의 똑같아보여서..
완벽하게 똑같은 `ProductInfoProductResponse`를 만드는 점이 굉장히 애매했습니다.
따라서 지금은 `ProductInfoAPI`에서 `HomeAPI`의 Dependency를 추가하여 `ProductResponse`를 끌어다(?) 쓰고 있습니다.

➡️ 그래서 이것은 Scene별로 나누지 않고, 서버에서 제시한 `EndPoint`로 나누는건 어떨까요? (`User`, `Search`, `Product`)

<br/><br/>

## References 📋


<br/><br/>

---

- Closed: #40

EnnzoKim committed:
📦️ ProductInfoAPI Package 디렉토리 추가

EnnzoKim committed:
✨ ProductInfoEndPoint 추가 및 구현

EnnzoKim committed:
🌱 ProductInfoPriceResponse DTO 구현

EnnzoKim committed:
🌱 Mock 데이터 추가 및 패키지 설정

EnnzoKim committed:
🌱 ProductInfoURLProtocol 구현

EnnzoKim committed:
🌱 ProductPrice Entity 추가

EnnzoKim committed:
✨ ProductInfoService 구현

EnnzoKim committed:
🎨 SwiftFormat 형식으로 코드 컨벤션 정리

EnnzoKim committed:
🌱 Product.init(dto:) private 해제 및 파일 이동

WhiteHyun commented on APIService/Sources/ProductInfoAPISupport/Mocks/ProductPriceResponse.json:
P2: 제가 깜빡하고 말씀 안드렸는데, date타입은 iso8601로 진행될 예정이라서 date타입도 그에 맞게 설정하시면 좋겠습니다 :)

- https://ko.wikipedia.org/wiki/ISO_8601

WhiteHyun commented on APIService/Sources/ProductInfoAPI/Responses/ProductPriceResponse.swift:
```diff
@@ -0,0 +1,26 @@
+    let dateString = try container.decode(Date.self, forKey: .date).formatted()
+    date = dateString
+    price = try container.decode(Int.self, forKey: .price)
+  }
```
P3: 이미 Provider에서 date 타입을 iso8601로 디코딩하고 있으므로 `decoder initializer`는 필요하지 않을 것 같아요. date타입도 `Date` 타입 그대로 받아도 괜찮을 것 같습니다 :)

WhiteHyun commented on APIService/Sources/ProductInfoAPI/ProductInfoService.swift:
```diff
@@ -0,0 +1,50 @@
+
+import Entity
+import Foundation
+import HomeAPI
```
P2: 같은 계층의 모듈끼리 참조하는 것은 위험성이 짙어요. 만약 상호참조하게 된다면 오류가 발생하게 됩니다. 꼭 분리해서 작업하는 것이 좋을 것 같다고 생각합니다.

WhiteHyun reviewed (commented):
제가 생각하기에는 `ProductInfo` 화면에서 카테고리(e.g. `음료 > 콜라`)도 보여줘야하는 것으로 알고 있어서 **나누는 게 맞다고 생각합니다.** 지금 2.0버전에서는 지원하지 않겠지만, 디자인에서는 카테고리가 존재하니까요..!
그래서 HomeService에서 사용되고 있는 API와 상세 정보를 가져오는 API가 지금으로써는 동일할지도 모르나 나중에는 결국 분리할 수 밖에 없을 거에요.

그리고 서버 EndPoint기준으로 Service를 구성하게 된다면, 나중에 Service를 주입할 때, **Service 배열**을 주입해야하는 경우가 생길 수 있고, 관리포인트가 늘어난다고 생각해서 화면마다 작업하는 게 좋다고 생각하고 있습니다. :)

EnnzoKim committed:
🌱 요청 사항 수정

- ProductDetailResponse DTO 추가
- 같은 모듈내에서 참조하는 것을 방지

EnnzoKim committed:
🌱 요청 사항 수정 및 날짜 변환 코드 추가

- init(decoder:) 제거
- ProductPrice init(dto:) 날짜 변환 코드 추가

EnnzoKim commented:
아! 카테고리가 있었군요 ㅎㅎ. 그런 맥락이라면 지금 컨벤션이 맞다고 생각됩니다!
요청 사항 모두 수정되었습니다.
+ 추가적으로 Entity의 `ProductPrice`를 DTO로 부터 변환할 때, 바로 `dateString`으로 변환되는 코드를 추가했습니다.

WhiteHyun commented on APIService/Sources/ProductInfoAPISupport/Mocks/ProductPriceResponse.json:
P2: 으잉? iSO8601 형식으로 설정하지 않으신 것 같아요. 백엔드에서는 시간을 보내줄 때 iso8601형식으로 보내주므로, 저희가 모킹할 json값도 iso8601에 맞게 설정하는 게 올바르지 않을까요?
참고로 iso8601 형식은 `2024-02-06T00:42:21Z`입니다.

EnnzoKim replied:
앗 죄송합니다! 파일이 많아서 완전히 까먹고 있었네요 🤣🤣

WhiteHyun commented on APIService/Sources/HomeAPI/Responses/ProductResponse.swift:
```diff
@@ -8,7 +8,9 @@
 import Foundation
 
-struct ProductResponse: Decodable {
+// MARK: - ProductResponse
+
+public struct ProductResponse: Decodable {
```
P2: `HomeAPI` 모듈 내에서 사용하므로 `public`은 지워주시면 좋겠습니다 ☺️

WhiteHyun commented on APIService/Sources/ProductInfoAPI/ProductInfoEndPoint.swift:
```diff
@@ -0,0 +1,41 @@
+// MARK: EndPoint
+
+extension ProductInfoEndPoint: EndPoint {
+  public var method: Network.HTTPMethod {
```
P3: 자동완성으로 생긴 것 같은데, `Network.HTTPMethod`를 `HTTPMethod`로 줄일 수 있을 것 같아요.

WhiteHyun commented on APIService/Sources/ProductInfoAPI/ProductInfoEndPoint.swift:
```diff
@@ -0,0 +1,41 @@
+    }
+  }
+
+  public var parameters: Network.HTTPParameter {
```
P3: 이하 동문입니다. 그냥 컨벤션 스타일이라서 무시하셔도 됩니다 ㅎㅎ

WhiteHyun commented on Entity/Sources/Entity/ProductPrice.swift:
```diff
@@ -0,0 +1,21 @@
+import Foundation
+
+public struct ProductPrice {
+  let date: String
```
P3: 이름은 `date`인데 타입은 `Date`가 아니네요, 의도한 건지 궁금합니다.

EnnzoKim replied:
네이밍이 `date`인 것은 딱히 의도한 건 아닙니다!
타입에 맞게 `dateString`으로 고치는 것이 나을까요?
++ 일단 2024.02에 맞게 `yearMonth`로 고쳤습니다!

WhiteHyun replied:
아래 리뷰를 참고하시면 될 것 같습니다 :-)

WhiteHyun commented on Entity/Sources/Entity/ProductPrice.swift:
```diff
@@ -0,0 +1,21 @@
+import Foundation
+
+public struct ProductPrice {
+  let date: String
+  let price: Int
+
+  public init(
+    date: String,
+    price: Int
+  ) {
+    self.date = date
+    self.price = price
+  }
```
P3: Entity 모듈에서만 `ProductPrice`의 프로퍼티가 사용되는 것이 아니라 다른 모듈에서도 사용한다면, 프로퍼티도 `public`을 붙이셔야 해요.

WhiteHyun commented on APIService/Sources/ProductInfoAPI/ProductInfoService.swift:
```diff
@@ -0,0 +1,69 @@
+  }
+}
+
+private extension Product {
+  init(dto: ProductDetailResponse) {
+    self.init(
+      id: dto.id,
+      imageURL: dto.imageURL,
+      price: dto.price,
+      name: dto.name,
+      promotion: dto.promotion,
+      convenienceStore: dto.convenienceStore
+    )
+  }
+}
```
P3: 지금은 `Product`를 사용해도 괜찮아보이나, 결국 분리될 존재라면 나중을 위해서 `ProductDetail`(이름 대충 지음)과 같이 새로운 모델로 처리해도 좋을 것 같아요.

EnnzoKim committed:
🌱 요청 사항 수정

- ProductPrice, yearMonth으로 변수 네이밍 변경
- 불필요한 접근 제어 제거
- 불필요한 코드 간략화
- ProductDetail, Entity 추가 및 적용
- ResponseDTO 날짜 값 iso8601으로 변경

WhiteHyun commented on APIService/Sources/ProductInfoAPI/ProductInfoService.swift:
```diff
@@ -0,0 +1,69 @@
+  }
+}
+
+private extension ProductPrice {
+  init(dto: ProductPriceResponse) {
+    let formatter = DateFormatter()
+    formatter.dateFormat = "yyyy.MM"
+    let yearMonth = formatter.string(from: dto.date)
+
+    self.init(
+      yearMonth: yearMonth,
+      price: dto.price
+    )
+  }
+}
```
P2: 모델 내에서 DateFormatter를 연산하는 것이 올바른지 고민해볼 필요가 있다고 생각해요.
저는 모델을 가지고 로직을 처리하는 것은 모델이 하는 역할이 아니라고 생각합니다.
그래서 `yyyy.MM`타입을 강제하기 보다, `Date`타입을 유지하되 모델을 사용하는 곳(e.g. `ViewModel`)에서 포맷팅을 처리하는 건 어떻게 보시나요?

EnnzoKim replied:
이런 사소한 문제지만 역할을 확실히 분리하는 것은 잘 신경쓰지 못 했는데, 감사합니다. 😊
수정되었습니다 ! !

EnnzoKim committed:
🌱 요청 사항 수정

- 이니셜라이저에 불필요한 코드 제거

EnnzoKim commented:
수정되었습니다.
포메팅 관련된 로직은 다음 PR때 `ViewModel`에서 구현하도록 하겠습니다. 👍

WhiteHyun commented on APIService/Sources/ProductInfoAPI/Responses/ProductPriceResponse.swift:
```diff
@@ -0,0 +1,13 @@
+
+import Foundation
+
+public struct ProductPriceResponse: Decodable {
```
P2: 이하동문입니다.

WhiteHyun commented on APIService/Sources/ProductInfoAPI/Responses/ProductDetailResponse.swift:
```diff
@@ -0,0 +1,27 @@
+import Entity
+import Foundation
+
+public struct ProductDetailResponse: Decodable {
```
P2: Network Model을 외부 모듈에서 사용하는 경우가 있을까요?
이미 `ProductDetail` Entity가 있는 것으로 보아 public은 불필요해 보여요.

EnnzoKim committed:
🌱 요청 사항 수정

- 불필요한 접근제어 삭제

WhiteHyun reviewed (approved):
고생하셨습니다 :)
